packages/react-query/src/__tests__/prefetch.test.tsx

```diff
@@ -0,0 +1,352 @@
+      await sleep(10)
+
+      throw new Error('Oops! Server error!')
+    })
+
+    function App() {
+      usePrefetchQuery(queryOpts)
+
+      return (
+        <ErrorBoundary fallbackRender={() => <div>Oops!</div>}>
+          <React.Suspense fallback="Loading...">
+            <Suspended queryOpts={queryOpts} />
+          </React.Suspense>
+        </ErrorBoundary>
+      )
+    }
+
+    await queryClient.prefetchQuery(queryOpts)
+    queryFn.mockClear()
+    const rendered = renderWithClient(queryClient, <App />)
+
+    await waitFor(() => rendered.getByText('Oops!'))
+    expect(rendered.queryByText('data: Not an error')).not.toBeInTheDocument()
+    expect(queryOpts.queryFn).not.toHaveBeenCalled()
+  })
+
+  it('should be able to recover from errors and try fetching again', async () => {
+    const queryFn = generateQueryFn('This is fine :dog: :fire:')
+
+    const queryOpts = {
+      queryKey: queryKey(),
+      queryFn,
+    }
+
+    queryFn.mockImplementationOnce(async () => {
+      await sleep(10)
+
+      throw new Error('Oops! Server error!')
+    })
+
+    function App() {
+      const { reset } = useQueryErrorResetBoundary()
+      usePrefetchQuery(queryOpts)
+
+      return (
+        <ErrorBoundary
+          onReset={reset}
+          fallbackRender={({ resetErrorBoundary }) => (
+            <div>
+              <div>Oops!</div>
+              <button onClick={resetErrorBoundary}>Try again</button>
+            </div>
+          )}
+        >
+          <React.Suspense fallback="Loading...">
+            <Suspended queryOpts={queryOpts} />
+          </React.Suspense>
+        </ErrorBoundary>
+      )
+    }
+
+    await queryClient.prefetchQuery(queryOpts)
+    queryFn.mockClear()
+
+    const rendered = renderWithClient(queryClient, <App />)
+
+    await waitFor(() => rendered.getByText('Oops!'))
+    fireEvent.click(rendered.getByText('Try again'))
+    await waitFor(() => rendered.getByText('data: This is fine :dog: :fire:'))
+    expect(queryOpts.queryFn).toHaveBeenCalledTimes(1)
+  })
+
+  it('should not create a suspense waterfall if prefetch is fired', async () => {
+    const firstQueryOpts = {
+      queryKey: queryKey(),
+      queryFn: generateQueryFn('Prefetch is nice!'),
+    }
+
+    const secondQueryOpts = {
+      queryKey: queryKey(),
+      queryFn: generateQueryFn('Prefetch is really nice!!'),
+    }
+
+    const thirdQueryOpts = {
+      queryKey: queryKey(),
+      queryFn: generateQueryFn('Prefetch does not create waterfalls!!'),
+    }
+
+    const Fallback = vi.fn().mockImplementation(() => <div>Loading...</div>)
+
+    function App() {
+      usePrefetchQuery(firstQueryOpts)
+      usePrefetchQuery(secondQueryOpts)
+      usePrefetchQuery(thirdQueryOpts)
+
+      return (
+        <React.Suspense fallback={<Fallback />}>
+          <Suspended queryOpts={firstQueryOpts} />
+          <Suspended queryOpts={secondQueryOpts} />
+          <Suspended queryOpts={thirdQueryOpts} />
```

Comment:
With sibling prerendering (the React 18 behaviour), this wouldn't waterfall even without the prefetch I think? If the `<Suspended>` component was remade to optionally take children, I think this would be a better test:

```suggestion
          <Suspended queryOpts={firstQueryOpts}>
            <Suspended queryOpts={secondQueryOpts}>
              <Suspended queryOpts={thirdQueryOpts} />
            </Suspended>
          </Suspended>
```